test/binder-call/binder-call.c

```diff
@@ -0,0 +1,726 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <gbinder.h>
+#include <gutil_log.h>
+
+#include <unistd.h>
+#include <assert.h>
+#include <binder-call.h>
+
+struct transaction_and_reply *ast;
+
+#define RET_OK          (0)
+#define RET_NOTFOUND    (1)
+#define RET_INVARG      (2)
+#define RET_ERR         (3)
+
+#define DEFAULT_DEVICE     GBINDER_DEFAULT_BINDER
+
+#define GBINDER_TRANSACTION(c2,c3,c4)     GBINDER_FOURCC('_',c2,c3,c4)
+#define GBINDER_INTERFACE_TRANSACTION     GBINDER_TRANSACTION('N','T','F')
+
+static const char pname[] = "binder-call";
+
+enum transaction_pass {
+    COMPUTE_SIZES = 0,
+    FILL_BUFFERS,
+    BUILD_TRANSACTION
+};
+
+enum reply_pass {
+    PRINT_REPLY = 0,
+    COMPUTE_SIZES_REPLY
+};
+
+int
+go_through_transaction_ast(
+    App *app,
+    GList *node_list,
+    int parent_idx,
+    void *buf,
+    enum transaction_pass cur_pass,
+    int cont_offset)
+{
+    int l = g_list_length(node_list);
+    int offset = cont_offset;
+
+    for (int i = 0; i < l; i++) {
+        struct value_info *v = g_list_nth_data(node_list, i);
+
+        switch(v->type) {
+            case INT8_TYPE:
+                if (cur_pass == BUILD_TRANSACTION) GDEBUG("int8");
+                if (parent_idx == -1) {
+                    gbinder_writer_append_int32(&app->writer, *((int*)v->value));
+                } else if (cur_pass == FILL_BUFFERS) {
+                    assert(buf != NULL);
+                    *((unsigned char*)(((char*)buf)+offset)) = *((unsigned char*)v->value);
+                }
+                offset += sizeof(int8_t);
+                break;
+            case INT32_TYPE:
+                if (cur_pass == BUILD_TRANSACTION) GDEBUG("int32");
+                if (parent_idx == -1) {
+                    gbinder_writer_append_int32(&app->writer, *((int*)v->value));
+                } else if (cur_pass == FILL_BUFFERS) {
+                    assert(buf != NULL);
+                    *((int*)(((char*)buf)+offset)) = *((int*)v->value);
+                }
+                offset += sizeof(int32_t);
+                break;
+            case INT64_TYPE:
+                if (cur_pass == BUILD_TRANSACTION) GDEBUG("int64");
+                if (parent_idx == -1) {
+                    gbinder_writer_append_int64(&app->writer, *((long*)v->value));
+                } else if (cur_pass == FILL_BUFFERS) {
+                    assert(buf != NULL);
+                    *((long*)(((char*)buf)+offset)) = *((long*)v->value);
+                }
+                offset += sizeof(int64_t);
+                break;
+            case FLOAT_TYPE:
+                if (cur_pass == BUILD_TRANSACTION) GDEBUG("float");
+                if (parent_idx == -1) {
+                    gbinder_writer_append_float(&app->writer, *((float*)v->value));
+                } else if (cur_pass == FILL_BUFFERS) {
```

Comment:
The same if under each case kind of suggests that two different things are going on here, and perhaps those two different things belong to two different functions... or maybe not